File: pkg/lib/html2po.py
#!/usr/bin/env python3
# This file is part of Cockpit.
#
# Copyright (C) 2023 Red Hat, Inc.
#
# Cockpit is free software; you can redistribute it and/or modify it
# under the terms of the GNU Lesser General Public License as published by
# the Free Software Foundation; either version 2.1 of the License, or
# (at your option) any later version.
#
# Cockpit is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with Cockpit; If not, see <http://www.gnu.org/licenses/>.

# Extracts translatable strings from HTML files in the following forms:
#
# <tag translate>String</tag>
# <tag translate context="value">String</tag>
# <tag translate="...">String</tag>
# <tag translate-attr attr="String"></tag>
#
# Supports the following angular-gettext compatible forms:
#
# <translate>String</translate>
# <tag translate-plural="Plural">Singular</tag>
#
# Note that some of the use of the translated may not support all the strings
# depending on the code actually using these strings to translate the HTML.

import argparse
import os
import sys
from typing import Dict

import bs4
import polib


def scan_tags(filename, entries, soup):
    """
    Look for tags marked as translatable and extract them
    """

    def is_translatable(tag):
        return tag.has_attr("translate") or tag.has_attr("translatable")

    tags = soup.find_all(is_translatable)

    # Extract translate strings
    for tag in tags:
        tasks = (
            tag.attrs.get("translate") or tag.attrs.get("translatable") or "yes"
        ).split(" ")

        # Calculate the line location
        line = tag.sourceline or 0

        msgctxt = tag.attrs.get("translate-context") or tag.attrs.get("context")
        msgid_plural = tag.attrs.get("translate-plural")
        location = (filename, str(line))

        # For each thing listed
        for task in tasks:
            msgid = None
            # The element text itself
            if task == "yes" or task == "translate":
                msgid = tag.get_text(strip=True)
            # An attribute
            elif task:
                msgid = tag.attrs[task]

            if msgid:
                push(
                    entries,
                    msgid=msgid,
                    msgid_plural=msgid_plural,
                    msgctxt=msgctxt,
                    location=location,
                )


def push(entries, msgid, msgid_plural, msgctxt, location):
    """
    Push an entry onto the list
    """
    key = msgid + "\0" + (msgid_plural or "") + "\0" + (msgctxt or "")
    if key in entries:
        entries[key].occurrences.append(location)
    else:
        entries[key] = polib.POEntry(
            msgid=msgid,
            msgid_plural=msgid_plural,
            msgctxt=msgctxt,
            occurrences=[location],
        )


def main() -> None:
    parser = argparse.ArgumentParser(
        prog="html2po",
        description="Extracts translatable strings from HTML files",
    )
    parser.add_argument("-d", "--directory", help="Base directory for input files")
    parser.add_argument("-o", "--output", help="Output file", required=True)
    parser.add_argument(
        "files", nargs="+", help="One or more input files", metavar="FILE"
    )

    args = parser.parse_args()

    # The unique messages extracted from the files
    entries: Dict[str, polib.POEntry] = {}

    # Now process each file in turn
    for filename in args.files:
        # Qualify the filename if necessary
        full = filename
        if args.directory:
            full = os.path.join(args.directory, filename)

        with open(full) as f:
            soup = bs4.BeautifulSoup(f, "html.parser")
        scan_tags(filename, entries, soup)

    po = polib.POFile()
    po.metadata = {
        "Project-Id-Version": "PACKAGE_VERSION",
        "MIME-Version": "1.0",
        "Content-Type": "text/plain; charset=UTF-8",
        "Content-Transfer-Encoding": "8bit",
        "X-Generator": "Cockpit html2po",
    }
    for entry in entries.values():
        po.append(entry)

    if args.output:
        po.save(args.output)
    else:
        sys.stdout.write(str(po))


if __name__ == "__main__":
    main()

File: pkg/lib/html2po.js
#!/usr/bin/env node

/*
 * Extracts translatable strings from HTML files in the following forms:
 *
 * <tag translate>String</tag>
 * <tag translate context="value">String</tag>
 * <tag translate="...">String</tag>
 * <tag translate-attr attr="String"></tag>
 *
 * Supports the following angular-gettext compatible forms:
 *
 * <translate>String</translate>
 * <tag translate-plural="Plural">Singular</tag>
 *
 * Note that some of the use of the translated may not support all the strings
 * depending on the code actually using these strings to translate the HTML.
 */

import fs from 'fs';
import path from 'path';
import htmlparser from 'htmlparser';
import { ArgumentParser } from 'argparse';

function fatal(message, code) {
    console.log((filename || "html2po") + ": " + message);
    process.exit(code || 1);
}

const parser = new ArgumentParser();
parser.add_argument('-d', '--directory', { help: "Base directory for input files" });
parser.add_argument('-o', '--output', { help: 'Output file', required: true });
parser.add_argument('files', { nargs: '+', help: "One or more input files", metavar: "FILE" });
const args = parser.parse_args();

const input = args.files;
const entries = { };

/* Filename being parsed and offset of line number */
let filename = null;
let offsets = 0;

/* The HTML parser we're using */
const handler = new htmlparser.DefaultHandler(function(error, dom) {
    if (error)
        fatal(error);
    else
        walk(dom);
});

/* Now process each file in turn */
step();

function step() {
    filename = input.shift();
    if (filename === undefined) {
        finish();
        return;
    }

    /* Qualify the filename if necessary */
    let full = filename;
    if (args.directory)
        full = path.join(args.directory, filename);

    fs.readFile(full, { encoding: "utf-8" }, function(err, data) {
        if (err)
            fatal(err.message);

        const parser = new htmlparser.Parser(handler, { includeLocation: true });
        parser.parseComplete(data);
        step();
    });
}

/* Process an array of nodes */
function walk(children) {
    if (!children)
        return;

    children.forEach(function(child) {
        const line = (child.location || { }).line || 0;
        const offset = line - 1;

        /* Scripts get their text processed as HTML */
        if (child.type == 'script' && child.children) {
            const parser = new htmlparser.Parser(handler, { includeLocation: true });

            /* Make note of how far into the outer HTML file we are */
            offsets += offset;

            child.children.forEach(function(node) {
                parser.parseChunk(node.raw);
            });
            parser.done();

            offsets -= offset;

        /* Tags get extracted as usual */
        } else if (child.type == 'tag') {
            tag(child);
        }
    });
}

/* Process a single loaded tag */
function tag(node) {
    let tasks, line, entry;
    const attrs = node.attribs || { };
    let nest = true;

    /* Extract translate strings */
    if ("translate" in attrs || "translatable" in attrs) {
        tasks = (attrs.translate || attrs.translatable || "yes").split(" ");

        /* Calculate the line location taking into account nested parsing */
        line = (node.location || { }).line || 0;
        line += offsets;

        entry = {
            msgctxt: attrs['translate-context'] || attrs.context,
            msgid_plural: attrs['translate-plural'],
            locations: [filename + ":" + line]
        };

        /* For each thing listed */
        tasks.forEach(function(task) {
            const copy = Object.assign({}, entry);

            /* The element text itself */
            if (task == "yes" || task == "translate") {
                copy.msgid = extract(node.children);
                nest = false;

            /* An attribute */
            } else if (task) {
                copy.msgid = attrs[task];
            }

            if (copy.msgid)
                push(copy);
        });
    }

    /* Walk through all the children */
    if (nest)
        walk(node.children);
}

/* Push an entry onto the list */
function push(entry) {
    const key = entry.msgid + "\0" + entry.msgid_plural + "\0" + entry.msgctxt;
    const prev = entries[key];
    if (prev) {
        prev.locations = prev.locations.concat(entry.locations);
    } else {
        entries[key] = entry;
    }
}

/* Extract the given text */
function extract(children) {
    if (!children)
        return null;

    const str = [];
    children.forEach(function(node) {
        if (node.type == 'tag' && node.children)
            str.push(extract(node.children));
        else if (node.type == 'text' && node.data)
            str.push(node.data);
    });

    const msgid = str.join("");

    if (msgid != msgid.trim()) {
        console.error("FATAL: string contains leading or trailing whitespace:", msgid);
        process.exit(1);
    }

    return msgid;
}

/* Escape a string for inclusion in po file */
function escape(string) {
    const bs = string.split('\\')
            .join('\\\\')
            .split('"')
            .join('\\"');
    return bs.split("\n").map(function(line) {
        return '"' + line + '"';
    }).join("\n");
}

/* Finish by writing out the strings */
function finish() {
    const result = [
        'msgid ""',
        'msgstr ""',
        '"Project-Id-Version: PACKAGE_VERSION\\n"',
        '"MIME-Version: 1.0\\n"',
        '"Content-Type: text/plain; charset=UTF-8\\n"',
        '"Content-Transfer-Encoding: 8bit\\n"',
        '"X-Generator: Cockpit html2po\\n"',
        '',
    ];

    for (const msgid in entries) {
        const entry = entries[msgid];
        result.push('#: ' + entry.locations.join(" "));
        if (entry.msgctxt)
            result.push('msgctxt ' + escape(entry.msgctxt));
        result.push('msgid ' + escape(entry.msgid));
        if (entry.msgid_plural) {
            result.push('msgid_plural ' + escape(entry.msgid_plural));
            result.push('msgstr[0] ""');
            result.push('msgstr[1] ""');
        } else {
            result.push('msgstr ""');
        }
        result.push('');
    }

    const data = result.join('\n');
    if (!args.output) {
        process.stdout.write(data);
        process.exit(0);
    } else {
        fs.writeFile(args.output, data, function(err) {
            if (err)
                fatal(err.message);
            process.exit(0);
        });
    }
}
